# 📡 Telecom Customer Churn Analytics

## End-to-End Data Analytics Project

### Project Overview

Customer churn is one of the biggest challenges faced by telecom companies. Losing existing customers leads to reduced revenue, increased customer acquisition costs, and lower profitability.

This project aims to analyze customer churn by combining traditional data analytics techniques with a modern ETL pipeline. The project follows the complete analytics lifecycle, including business understanding, data cleaning, exploratory data analysis (EDA), feature engineering, AWS S3 integration, PostgreSQL data warehousing (Bronze → Silver → Gold), SQL business analysis, and interactive Power BI dashboards.

The objective is to identify the major factors contributing to customer churn and provide actionable business recommendations that help improve customer retention.

# 🎯 Business Problem

A telecom company has observed an increasing number of customers discontinuing its services. Customer churn directly impacts revenue, customer lifetime value, and long-term business growth.

The company wants to better understand customer behavior and identify the key reasons why customers leave. By analyzing customer demographics, subscribed services, billing information, contract details, and payment methods, the business aims to discover patterns associated with customer churn.

The insights generated from this analysis will help management design strategies to improve customer retention and reduce revenue loss.

# 🎯 Business Objectives

The primary objectives of this project are:

- Understand customer churn behavior.
- Identify the major factors influencing customer churn.
- Analyze customer demographics and service usage.
- Measure customer retention and churn rate.
- Analyze revenue generated by different customer segments.
- Identify high-risk customer groups.
- Build business-ready KPIs for management.
- Develop interactive Power BI dashboards.
- Provide actionable business recommendations for improving customer retention.

# ❓ Business Questions

This project aims to answer the following business questions:

1. What is the overall customer churn rate?

2. Which contract type experiences the highest customer churn?

3. Which payment method is associated with the highest churn?

4. Does internet service type influence customer churn?

5. Which customer segment contributes the highest revenue?

6. Does customer tenure affect churn?

7. How do monthly charges influence customer churn?

8. Which value-added services reduce customer churn?

9. Which customer segments should the company focus on retaining?

10. What business strategies can help reduce customer churn?

# 📊 Key Performance Indicators (KPIs)

The following KPIs will be used throughout the project:

- Total Customers
- Churn Customers
- Active Customers
- Customer Churn Rate
- Customer Retention Rate
- Total Monthly Revenue
- Average Monthly Charges
- Average Customer Tenure
- Revenue Lost due to Churn
- Average Revenue per Customer

# 📂 Dataset Understanding

This project uses the IBM Telco Customer Churn Dataset.

The dataset contains customer demographic information, subscribed telecom services, contract details, billing information, payment methods, and churn status.

The target variable is **Churn**, indicating whether a customer has left the company.

Each row represents one customer, while each column represents a customer attribute or subscribed service.

### Data Inspection

In [2]:
import pandas as pd

In [3]:
#load dataset 
df = pd.read_csv("../data/raw/WA_Fn-UseC_-Telco-Customer-Churn.csv")

In [4]:
#display first 5 rows
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


# 📋 Dataset Summary

We will examine:

- Number of rows and columns
- Data types
- Missing values
- Duplicate records
- Statistical summary
- Categorical variables
- Numerical variables

This helps us understand the quality and structure of the dataset before performing data cleaning.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


-- null values present in dataset 

In [6]:
df.describe().round()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.0,7043.0,7043.0
mean,0.0,32.0,65.0
std,0.0,25.0,30.0
min,0.0,0.0,18.0
25%,0.0,9.0,36.0
50%,0.0,29.0,70.0
75%,0.0,55.0,90.0
max,1.0,72.0,119.0


-- it does not show any outliers 

In [7]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

-- No Null values present in dataset

In [8]:
df.shape

(7043, 21)

-- it shows that total 7043 rows and 21 columns are present

In [9]:
df[df.duplicated()]

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn


In [10]:
df[df.duplicated()].shape

(0, 21)

In [11]:
df['customerID'].value_counts().head()

customerID
3186-AJIEK    1
7590-VHVEG    1
5575-GNVDE    1
8775-CEBBJ    1
2823-LKABH    1
Name: count, dtype: int64

In [12]:
df[df['customerID'] == '3186-AJIEK']

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,...,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


--- No duplicates are present in dataset

In [13]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [14]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

#### Inspecting some columns to see which type of data is present

In [15]:
df['gender'].unique()

array(['Female', 'Male'], dtype=object)

In [16]:
df['SeniorCitizen'].unique()

array([0, 1])

In [17]:
df['Partner'].unique()

array(['Yes', 'No'], dtype=object)

In [18]:
df['Dependents'].unique()

array(['No', 'Yes'], dtype=object)

In [19]:
df['tenure'].unique()

array([ 1, 34,  2, 45,  8, 22, 10, 28, 62, 13, 16, 58, 49, 25, 69, 52, 71,
       21, 12, 30, 47, 72, 17, 27,  5, 46, 11, 70, 63, 43, 15, 60, 18, 66,
        9,  3, 31, 50, 64, 56,  7, 42, 35, 48, 29, 65, 38, 68, 32, 55, 37,
       36, 41,  6,  4, 33, 67, 23, 57, 61, 14, 20, 53, 40, 59, 24, 44, 19,
       54, 51, 26,  0, 39])

In [20]:
df['OnlineBackup'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [21]:
df['OnlineSecurity'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [22]:
df['InternetService'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [23]:
df['Contract'].unique()

array(['Month-to-month', 'One year', 'Two year'], dtype=object)

In [24]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

In [25]:
df['StreamingMovies'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)